common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -22,64 +22,379 @@
+                                }
+                                request.cancel();
+                                Logger.infoPII(methodTag,  "User has reached the maximum failed attempts allowed.");
+                                return;
+                            }
+                            //Create ArrayList that contains cert details only pertinent to the cert picker.
+                            final ArrayList<YubiKitCertDetails> certList = new ArrayList<>();
+                            //We need to check every slot. If a slot is empty, an ApduException is called, in which case, we'll ignore it.
+                            //AUTHENTICATION (9A)
+                            getAndPutCertDetailsInList(AUTHENTICATION, piv, certList);
+                            //SIGNATURE (9C)
+                            getAndPutCertDetailsInList(SIGNATURE, piv, certList);
+                            //KEY_MANAGEMENT (9D)
+                            getAndPutCertDetailsInList(KEY_MANAGEMENT, piv, certList);
+                            //CARD_AUTH (9E)
+                            getAndPutCertDetailsInList(CARD_AUTH, piv, certList);
+                            //If no certs were found, cancel flow.
+                            if (certList.isEmpty()) {
+                                request.cancel();
+                                Logger.infoPII(methodTag,  "No PIV certificates found on YubiKey device.");
+                                return;
+                            }
+                            //Build and show Smartcard cert picker, which also handles the rest of the smartcard CBA flow.
+                            final SmartcardCertPickerDialog certPickerDialog = new SmartcardCertPickerDialog(certList, mActivity);
+                            certPickerDialog.setPositiveButtonListener(getSmartcardCertPickerDialogPositiveButtonListener(request));
+                            certPickerDialog.setNegativeButtonListener(new SmartcardCertPickerDialog.NegativeButtonListener() {
+                                @Override
+                                public void onClick() {
+                                    request.cancel();
+                                    mCurrentDialog = null;
+                                }
+                            });
+                            certPickerDialog.setCancelCbaCallback(new SmartcardCertPickerDialog.CancelCbaCallback() {
+                                @Override
+                                public void onCancel() {
+                                    request.cancel();
+                                }
+                            });
+
+                            certPickerDialog.show();
+                            //Set current dialog to certPickerDialog
+                            synchronized (smartcardDialogLock) {
+                                mCurrentDialog = certPickerDialog;
+                            }
+
```

Comment:
I wanted to put a `ResultFuture<YubiKitCertDetails>` `get()` call here, and then set the `setResult(null)` in the `onCancel()` method above. However, it seems that the `get()` statement is also blocking the `setOnClosed()` runnable that is run whenever the yubikey is unplugged. The result is that nothing in the `setOnClosed()` runnable is ever run.

Reply:
Not sure which line(s) are you referring to...

Reply:
Line 225